src/psij/executors/batch/batch_scheduler_executor.py

```diff
@@ -0,0 +1,539 @@
+    """
+
+    def __init__(self, url: Optional[str] = None,
+                 config: Optional[BatchSchedulerExecutorConfig] = None):
+        """Initializes a `BatchSchedulerExecutor`.
+
+        Parameters
+        ----------
+        url
+            An optional URL pointing to a specific backend
+        config
+            An configuration for this executor instance; if none is specified, a default
+            configuration is used.
+        """
+        super().__init__(url=url, config=config if config else BatchSchedulerExecutorConfig())
+        assert config
+        self.work_directory = config.work_directory / self.name
+        self._queue_poll_thread = self._start_queue_poll_thread()
+
+    def _ensure_work_dir(self) -> None:
+        self.work_directory.mkdir(parents=True, exist_ok=True)
+
+    def submit(self, job: Job) -> None:
+        """See :func:`~JobExecutor.submit`."""
+        logger.info('Job %s: submitting', job.id)
+        self._ensure_work_dir()
+        assert (job.spec)
+
+        context = self._create_script_context(job)
+
+        # assumes job ids are unique
+        submit_file_path = self.work_directory / (job.id + '.job')
+        with submit_file_path.open('w') as submit_file:
+            self.generate_submit_script(job, context, submit_file)
+        try:
+            logger.debug('Job %s: running submit command', job.id)
+            out = self._run_command(self.get_submit_command(job, submit_file_path))
+            logger.debug('Job %s: submit command ouput: %s', job.id, out)
+            job.native_id = self.job_id_from_submit_output(out)
+            logger.info('Job %s: native id: %s', job.id, job.native_id)
+            self._set_job_status(job, JobStatus(JobState.QUEUED,
+                                                metadata={'native_id': job.native_id}))
+        except subprocess.CalledProcessError as ex:
+            raise SubmitException(ex.output) from None
+
+        self._queue_poll_thread.register_job(job)
+
+    def _get_launcher_from_job(self, job: Job) -> Launcher:
+        assert(job.spec)
+        launcher_name = job.spec.launcher
+        if not launcher_name:
+            launcher_name = Launcher.DEFAULT_LAUNCHER_NAME
+        return self._get_launcher(launcher_name)
+
+    def cancel(self, job: Job) -> None:
+        """Cancels a job if it has not otherwise completed.
+
+        A command is constructed using :func:`~get_cancel_command` and executed in order to cancel
+        the job. Also see :func:`~JobExecutor.cancel`.
+        """
+        if job.native_id is None:
+            raise SubmitException('Job does not have a native ID.')
+        if job.status.state.final:
+            return
+        try:
+            self._run_command(self.get_cancel_command(job))
+        except subprocess.CalledProcessError as ex:
+            if job.status.state.final and not job.status.state == JobState.CANCELED:
```

Comment:
That's true. This is a pretty tricky part. The best solution would be for the LRM to properly report this situation and for the executor to understand LRM error messages. But this can't be implemented at the base class level. 

Suggestions welcome.

Reply:
In the except clause, one could call
```py
out = self.executor._run_command(self.executor.get_status_command(job.native_id))
[...]
```
to explicitly and synchronously check the job state.  That probably would imply some refactoring to avoid code repetition with the `_poll` thread.  But even that is not really save: batch systems don't always report state changes timely, and the state check could still signal a conflict.  I am not sure how one would handle that (although, that situation would be much rarer than the above which looks like could trigger frequently for purged job since the default poll frequency is 30 seconds - lots of time for this race to hit).

Reply:
I think the proper way would be to check the output of `scancel` / `qdel`. It almost surely will tell why it failed.